interfaces/builtin/kubernetes_support.go

```diff
@@ -35,65 +47,227 @@ const kubernetesSupportBaseDeclarationSlots = `
-/etc/modprobe.d/{,**} r,
+var kubernetesSupportConnectedPlugKmodKubeProxy = []string{
+	`ip_vs_rr`,
+	`ip_vs_sh`,
+	`ip_vs_wrr`,
+	`libcrc32c`,
+	`llc`,
+	`stp`,
+}
 
-# Allow ptracing Docker containers
-ptrace (read, trace) peer=docker-default,
-ptrace (read, trace) peer=snap.docker.dockerd,
+type kubernetesSupportInterface struct{}
 
-# Should we have a 'privileged' mode for kubernetes like we do with
-# docker-support? Right now kubernetes needs this which allows it to control
-# all processes on the system and on <4.8 kernels, escape confinement.
-ptrace (read, trace) peer=unconfined,
-`
+func (iface *kubernetesSupportInterface) Name() string {
+	return "kubernetes-support"
+}
+
+func (iface *kubernetesSupportInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              kubernetesSupportSummary,
+		BaseDeclarationPlugs: kubernetesSupportBaseDeclarationPlugs,
+		BaseDeclarationSlots: kubernetesSupportBaseDeclarationSlots,
+		ImplicitOnCore:       true,
+		ImplicitOnClassic:    true,
+	}
+}
+
+func (iface *kubernetesSupportInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	snippet := kubernetesSupportConnectedPlugAppArmorCommon
 
-var kubernetesSupportConnectedPlugKmod = []string{`llc`, `stp`}
+	systemd_run_extra := ""
+	var flavor string
+	_ = plug.Attr("flavor", &flavor)
+	if flavor == "kubelet" {
+		systemd_run_extra = kubernetesSupportConnectedPlugAppArmorKubeletSystemdRun
+		snippet += kubernetesSupportConnectedPlugAppArmorKubelet
+		spec.UsesPtraceTrace()
+	} else if flavor == "kubeproxy" {
+		snippet += kubernetesSupportConnectedPlugAppArmorKubeproxy
+	} else {
+		systemd_run_extra = kubernetesSupportConnectedPlugAppArmorKubeletSystemdRun
+		snippet += kubernetesSupportConnectedPlugAppArmorKubelet
+		snippet += kubernetesSupportConnectedPlugAppArmorKubeproxy
+		spec.UsesPtraceTrace()
+	}
+
+	old := "###KUBERNETES_SUPPORT_SYSTEMD_RUN###"
+	spec.AddSnippet(strings.Replace(snippet, old, systemd_run_extra, -1))
+	return nil
+}
+
+func (iface *kubernetesSupportInterface) SecCompConnectedPlug(spec *seccomp.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var flavor string
+	_ = plug.Attr("flavor", &flavor)
+
+	if flavor != "kubeproxy" {
+		snippet := kubernetesSupportConnectedPlugSeccompKubelet
+		spec.AddSnippet(snippet)
+	}
+	return nil
+}
+
+func (iface *kubernetesSupportInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var flavor string
+	_ = plug.Attr("flavor", &flavor)
+
+	if flavor != "kubeproxy" {
+		for _, rule := range kubernetesSupportConnectedPlugUDevKubelet {
+			spec.TagDevice(rule)
+		}
+	}
+	return nil
+}
+
+func (iface *kubernetesSupportInterface) KModConnectedPlug(spec *kmod.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var flavor string
```

Comment:
Since this pattern repeats often it might make sense to add a small helper:

```
func k8sFlavour(plug *interfaces.ConnectedPlug) string {
    var flavour string
    _ = plug.Attr("flavor", &flavour)
    return flavour
}
```

PS: please forgive the British English spelling above.

Reply:
Done